extensions/druid-rocketmq/src/main/java/io/druid/firehose/rocketmq/RocketMQFirehoseFactory.java

```diff
@@ -0,0 +1,571 @@
+              }
+            }
+            offsetStore.persistAll(updated);
+          }
+        };
+      }
+
+      @Override
+      public void close() throws IOException
+      {
+        defaultMQPullConsumer.shutdown();
+        pullMessageService.shutdown(false);
+      }
+    };
+  }
+
+
+  /**
+   * Pull request.
+   */
+  final class DruidPullRequest
+  {
+    private final MessageQueue messageQueue;
+    private final String tag;
+    private final long nextBeginOffset;
+    private final int pullBatchSize;
+    private final boolean longPull;
+    private final CountDownLatch countDownLatch;
+
+    public DruidPullRequest(
+        final MessageQueue messageQueue,
+        final String tag,
+        final long nextBeginOffset,
+        final int pullBatchSize,
+        final boolean useLongPull
+    )
+    {
+      this.messageQueue = messageQueue;
+      this.tag = (null == tag ? "*" : tag);
+      this.nextBeginOffset = nextBeginOffset;
+      this.pullBatchSize = pullBatchSize;
+      this.longPull = useLongPull;
+      countDownLatch = new CountDownLatch(1);
+    }
+
+    public MessageQueue getMessageQueue()
+    {
+      return messageQueue;
+    }
+
+    public long getNextBeginOffset()
+    {
+      return nextBeginOffset;
+    }
+
+    public String getTag()
+    {
+      return tag;
+    }
+
+    public int getPullBatchSize()
+    {
+      return pullBatchSize;
+    }
+
+    public boolean isLongPull()
+    {
+      return longPull;
+    }
+
+    public CountDownLatch getCountDownLatch()
+    {
+      return countDownLatch;
+    }
+  }
+
+
+  /**
+   * Pull message service for druid.
+   * <p/>
+   * <strong>Note: this is a single thread service.</strong>
+   */
+  final class DruidPullMessageService extends ServiceThread
+  {
+
+    private volatile List<DruidPullRequest> requestsWrite = new ArrayList<>();
+    private volatile List<DruidPullRequest> requestsRead = new ArrayList<>();
+
+    private final DefaultMQPullConsumer defaultMQPullConsumer;
+
+    public DruidPullMessageService(final DefaultMQPullConsumer defaultMQPullConsumer)
+    {
+      this.defaultMQPullConsumer = defaultMQPullConsumer;
+    }
+
+    public void putRequest(final DruidPullRequest request)
+    {
+      synchronized (this) {
+        this.requestsWrite.add(request);
+        if (!hasNotified) {
+          hasNotified = true;
+          notify();
+        }
+      }
+    }
+
+    private void swapRequests()
+    {
+      List<DruidPullRequest> tmp = requestsWrite;
+      requestsWrite = requestsRead;
+      requestsRead = tmp;
+    }
+
+    @Override
+    public String getServiceName()
+    {
+      return getClass().getSimpleName();
+    }
+
+    /**
+     * Core message pulling logic code goes here.
+     */
+    private void doPull()
+    {
+      for (DruidPullRequest pullRequest : requestsRead) {
+        PullResult pullResult;
+        try {
+          if (!pullRequest.isLongPull()) {
+            pullResult = defaultMQPullConsumer.pull(
```

Comment:
Default timeout looks to be 10 seconds, `DefaultMQPullConsumer` has lots of settings which are not tune-able in this extension. Is there any reasonable or standard-ish way to pull in configurations for `DefaultMQPullConsumer`?